File: backend/app/schemas/metadata.py
from datetime import datetime
from enum import Enum
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field

class ClientType(str, Enum):
    LOCAL_GOVERNMENT = "LOCAL_GOVERNMENT"       # 지방자치단체
    CENTRAL_GOVERNMENT = "CENTRAL_GOVERNMENT"   # 국가기관 / 중앙행정기관
    PUBLIC_INSTITUTION = "PUBLIC_INSTITUTION"   # 공공기관 / 공기업
    EDUCATIONAL = "EDUCATIONAL"                 # 교육청 / 국공립학교
    OTHER = "OTHER"                             # 기타

class GoverningLaw(str, Enum):
    LOCAL_CONTRACT_ACT = "LOCAL_CONTRACT_ACT"   # 지방계약법
    STATE_CONTRACT_ACT = "STATE_CONTRACT_ACT"   # 국가계약법
    PUBLIC_ENTERPRISE_RULE = "PUBLIC_ENTERPRISE_RULE" # 공기업·준정부기관 계약사무규칙
    OTHER = "OTHER"                             # 기타

class ProcurementMethod(str, Enum):
    NEGOTIATION = "NEGOTIATION"                 # 협상에 의한 계약
    RESTRICTED_COMPETITIVE = "RESTRICTED_COMPETITIVE" # 제한경쟁입찰
    OPEN_COMPETITIVE = "OPEN_COMPETITIVE"       # 일반경쟁입찰
    PRIVATE_CONTRACT = "PRIVATE_CONTRACT"       # 수의계약

class ExtractedMetadata(BaseModel):
    project_id: str = Field(..., description="연결된 프로젝트 ID")
    project_name: Optional[str] = Field(None, description="사업명")
    client_name: Optional[str] = Field(None, description="수요기관명")
    client_type: ClientType = Field(ClientType.LOCAL_GOVERNMENT, description="수요기관 유형")
    governing_law: GoverningLaw = Field(GoverningLaw.LOCAL_CONTRACT_ACT, description="적용 계약법")
    procurement_method: ProcurementMethod = Field(ProcurementMethod.NEGOTIATION, description="계약방법")
    budget_amount: Optional[int] = Field(None, description="사업예산 (원, 부가세 포함)")
    estimated_price: Optional[int] = Field(None, description="추정가격 (원, 부가세 제외)")
    project_period: Optional[str] = Field(None, description="사업기간")
    confidence_scores: Dict[str, float] = Field(default_factory=dict, description="필드별 AI 추출 신뢰도 (0.0~1.0)")
    source_references: Dict[str, str] = Field(default_factory=dict, description="필드별 추출 근거 block_id")
    extracted_at: datetime = Field(default_factory=datetime.utcnow, description="추출 일시")
    status: str = Field("COMPLETED", description="추출 상태 (COMPLETED, FAILED)")

class AuthoritativeMetadata(BaseModel):
    project_id: str = Field(..., description="연결된 프로젝트 ID")
    version: int = Field(1, ge=1, description="확정본 버전 (1부터 시작)")
    project_name: str = Field(..., description="사업명")
    client_name: str = Field(..., description="수요기관명")
    client_type: ClientType = Field(..., description="수요기관 유형")
    governing_law: GoverningLaw = Field(..., description="적용 계약법")
    procurement_method: ProcurementMethod = Field(..., description="계약방법")
    budget_amount: Optional[int] = Field(None, description="사업예산 (원, 부가세 포함)")
    estimated_price: Optional[int] = Field(None, description="추정가격 (원, 부가세 제외)")
    project_period: Optional[str] = Field(None, description="사업기간")
    confirmed_by: Optional[str] = Field("user_officer", description="확정자 식별자")
    note: Optional[str] = Field(None, description="확정 비고 및 메모")
    updated_at: datetime = Field(default_factory=datetime.utcnow, description="최종 확정 일시")

class AuthoritativeMetadataUpdateDto(BaseModel):
    project_name: str
    client_name: str
    client_type: ClientType
    governing_law: GoverningLaw
    procurement_method: ProcurementMethod
    budget_amount: Optional[int] = None
    estimated_price: Optional[int] = None
    project_period: Optional[str] = None
    confirmed_by: Optional[str] = "user_officer"
    note: Optional[str] = None

class MetadataSnapshot(BaseModel):
    snapshot_id: str
    project_id: str
    version: int
    data: AuthoritativeMetadata
    created_at: datetime = Field(default_factory=datetime.utcnow)

File: src/types/metadata.ts
export type ClientType =
  | 'LOCAL_GOVERNMENT'
  | 'CENTRAL_GOVERNMENT'
  | 'PUBLIC_INSTITUTION'
  | 'EDUCATIONAL'
  | 'OTHER';

export type GoverningLaw =
  | 'LOCAL_CONTRACT_ACT'
  | 'STATE_CONTRACT_ACT'
  | 'PUBLIC_ENTERPRISE_RULE'
  | 'OTHER';

export type ProcurementMethod =
  | 'NEGOTIATION'
  | 'RESTRICTED_COMPETITIVE'
  | 'OPEN_COMPETITIVE'
  | 'PRIVATE_CONTRACT';

export interface ExtractedMetadata {
  project_id: string;
  project_name?: string | null;
  client_name?: string | null;
  client_type: ClientType;
  governing_law: GoverningLaw;
  procurement_method: ProcurementMethod;
  procurement_method_reason?: string;
  extracted_snippets?: Record<string, string>;
  is_ai_powered?: boolean;
  budget_amount?: number | null;
  estimated_price?: number | null;
  project_period?: string | null;
  confidence_scores: Record<string, number>;
  source_references: Record<string, string>;
  extracted_at: string;
  status: string;
}

export interface AuthoritativeMetadata {
  project_id: string;
  version: number;
  project_name: string;
  client_name: string;
  client_type: ClientType;
  governing_law: GoverningLaw;
  procurement_method: ProcurementMethod;
  budget_amount?: number | null;
  estimated_price?: number | null;
  project_period?: string | null;
  confirmed_by?: string;
  note?: string | null;
  updated_at: string;
}

export interface AuthoritativeMetadataUpdateDto {
  project_name: string;
  client_name: string;
  client_type: ClientType;
  governing_law: GoverningLaw;
  procurement_method: ProcurementMethod;
  budget_amount?: number | null;
  estimated_price?: number | null;
  project_period?: string | null;
  confirmed_by?: string;
  note?: string | null;
}

export interface MetadataSnapshot {
  snapshot_id: string;
  project_id: string;
  version: number;
  data: AuthoritativeMetadata;
  created_at: string;
}

export const CLIENT_TYPE_LABELS: Record<ClientType, { label: string; desc: string }> = {
  LOCAL_GOVERNMENT: { label: '지방자치단체', desc: '특별시·광역시·도, 시·군·구 및 직속기관 (지방계약법 적용)' },
  CENTRAL_GOVERNMENT: { label: '국가기관 / 중앙행정기관', desc: '부·처·청 및 소속 국가기관 (국가계약법 적용)' },
  PUBLIC_INSTITUTION: { label: '공공기관 / 공기업', desc: '공공기관운영법 대상 공기업 및 준정부기관' },
  EDUCATIONAL: { label: '교육청 / 국공립학교', desc: '시·도 교육청 및 산하 교육지원청, 국공립학교' },
  OTHER: { label: '기타 공공단체', desc: '기타 특별법인 및 출연연구기관' },
};

export const GOVERNING_LAW_LABELS: Record<GoverningLaw, { label: string; short: string }> = {
  LOCAL_CONTRACT_ACT: { label: '지방자치단체를 당사자로 하는 계약에 관한 법률 (지방계약법)', short: '지방계약법' },
  STATE_CONTRACT_ACT: { label: '국가를 당사자로 하는 계약에 관한 법률 (국가계약법)', short: '국가계약법' },
  PUBLIC_ENTERPRISE_RULE: { label: '공기업·준정부기관 계약사무규칙', short: '공기업계약규칙' },
  OTHER: { label: '기타 관계 법령', short: '기타' },
};

export const PROCUREMENT_METHOD_LABELS: Record<ProcurementMethod, { label: string; desc: string }> = {
  NEGOTIATION: { label: '협상에 의한 계약', desc: '기술제안서 평가(80~90%) + 가격평가(10~20%)' },
  RESTRICTED_COMPETITIVE: { label: '제한경쟁입찰', desc: '지역, 실적, 자격요건 등을 제한하는 경쟁' },
  OPEN_COMPETITIVE: { label: '일반경쟁입찰', desc: '모든 자격보유자의 공개 입찰 참가 허용' },
  PRIVATE_CONTRACT: { label: '수의계약', desc: '소액 또는 특정 사유에 의한 특정업체 직접 계약' },
};
